app/src/main/java/com/example/dogadoption/repository/DogRepository.kt

```diff
@@ -4,41 +4,67 @@ import android.content.ContentValues
-        return withContext(Dispatchers.IO) {
-            dogApi.getDogBreeds().message.keys.toList()
-        }
+    fun getDogBreeds(): Flow<List<DogNames>> {
+        return localSource.getDogBreeds()
     }
 
-    suspend fun getDogPictures(breed: String): Result<List<String>> {
-        return withContext(Dispatchers.IO) {
-            runCatching {
-                val response = dogApi.getDogPictures(breed)
-                handleApiResponse(response)
+    fun getDogImages(breed: String): Flow<List<String>> {
+        Log.e("DOG_REPOSITORY", "Fetched dog images for $breed")
+        return localSource.getDogBreedImages(breed)
+
+    }
+
+    suspend fun fetchAndSaveDogBreeds() {
+        val dogBreeds = localSource.getDogBreeds().firstOrNull()
+        if (dogBreeds.isNullOrEmpty()) {
+            try {
+                val remoteDogBreeds = remoteSource.getDogBreeds()
+                val dogNamesList = remoteDogBreeds.map { DogNames(0, it, "") }
+                localSource.insertDogBreed(dogNamesList)
+            } catch (e: Exception) {
+                Log.e("DOG_REPOSITORY", "Failed to fetch dog breeds: ${e.message}")
+                throw e
             }
         }
     }
-    private fun handleApiResponse(response: Response<DogPictures>): List<String> {
-        if (response.isSuccessful) {
-            val dogImages = response.body()
-            if (dogImages != null && dogImages.status == "success") {
-                Log.e(ContentValues.TAG,"REPOSITORY: API response [SUCCESS!]: ${response.code()}")
-                return dogImages.message
-            } else {
-                Log.e(ContentValues.TAG,"REPOSITORY: Unexpected API response: ${response.code()}")
-                throw Exception("Unexpected API response: ${response.code()}")
+
+    suspend fun fetchAndSaveDogPictures(breed: String) {
+        val dogImageDB = localSource.getDogBreedImages(breed).firstOrNull()
+        if (dogImageDB.isNullOrEmpty()) {
+            try {
+                val result = remoteSource.getDogPictures(breed)
+                result.onSuccess { dogPictures ->
+                    dogPictures.forEach { imageUrl ->
+                        val existingImages = localSource.getDogBreedImages(breed).firstOrNull()
+                        if (existingImages == null || !existingImages.contains(imageUrl)) {
+                            val dogImages = DogImages(0, imageUrl, breed)
+                            localSource.insertDogBreedImages(dogImages)
+                        }
+                    }
+                    Log.d("DOG_REPOSITORY", "Dog pictures saved to database for $breed")
+                }.onFailure { e ->
+                    Log.e("DOG_REPOSITORY", "Failed to fetch dog pictures: ${e.message}")
+                    throw e
+                }
+            } catch (e: Exception) {
+                Log.e("DOG_REPOSITORY", "Failed to save dog pictures to database: ${e.message}")
+                throw e
```

Comment:
this is a bit messy. you have kotlin result, which you unwrap, then on failure, you throw, then you catch again, and you throw it again...